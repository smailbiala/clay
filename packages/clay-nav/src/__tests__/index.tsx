/**
 * SPDX-FileCopyrightText: Â© 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

import {ClayVerticalNav} from '..';
import {
	cleanup,
	fireEvent,
	getAllByRole as docGetAllByRole,
	render,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

const ITEMS = [
	{
		items: [
			{
				href: '#',
				label: 'Nested1',
			},
		],
		label: 'Home',
	},
	{
		href: '#',
		label: 'About',
	},
	{
		href: '#',
		label: 'Contact',
	},
	{
		items: [
			{
				active: true,
				href: '#',
				label: 'Five',
			},
			{
				href: '#',
				label: 'Six',
			},
		],
		label: 'Projects',
	},
	{
		href: '#',
		label: 'Seven',
	},
];

describe('ClayVerticalNav', () => {
	afterEach(() => cleanup());

	const items = [
		{
			initialExpanded: true,
			items: [
				{
					href: '#nested1',
					label: 'Nested1',
				},
			],
			label: 'Home',
		},
	];

	it('renders', () => {
		const {container} = render(
			<ClayVerticalNav items={items} spritemap="/path/to" />
		);

		expect(container).toMatchSnapshot();
	});

	it('expands items when clicked', () => {
		const {container, getByText} = render(
			<ClayVerticalNav items={items} spritemap="/path/to" />
		);

		expect(
			container.querySelector('.collapsing.show .nav-item')
		).toBeFalsy();

		fireEvent.click(getByText('Home'));

		expect(
			container.querySelector('.collapsing.show .nav-item')
		).toBeTruthy();
	});

	it('expand items by pressing the right arrow key', () => {
		const {getAllByRole} = render(
			<ClayVerticalNav items={ITEMS} spritemap="/path/to" />
		);

		const [, , projects] = getAllByRole('button');

		userEvent.tab();
		userEvent.tab();

		expect(document.activeElement).toEqual(projects);

		userEvent.keyboard('[ArrowRight]');

		expect(projects!.getAttribute('aria-expanded')).toBe('true');
	});

	it('collapse items by pressing the left arrow key', () => {
		const {getAllByRole} = render(
			<ClayVerticalNav items={ITEMS} spritemap="/path/to" />
		);

		const [, , projects] = getAllByRole('button');

		userEvent.tab();
		userEvent.tab();

		expect(document.activeElement).toEqual(projects);

		userEvent.keyboard('[ArrowRight]');

		expect(projects!.getAttribute('aria-expanded')).toBe('true');

		userEvent.keyboard('[ArrowLeft]');

		expect(projects!.getAttribute('aria-expanded')).toBe('false');
	});

	it('moves focus to first item if item is expanded', () => {
		const {getAllByRole} = render(
			<ClayVerticalNav items={ITEMS} spritemap="/path/to" />
		);

		const [, , projects] = getAllByRole('button');

		userEvent.tab();
		userEvent.tab();

		expect(document.activeElement).toEqual(projects);

		userEvent.keyboard('[ArrowRight]');

		expect(projects!.getAttribute('aria-expanded')).toBe('true');

		userEvent.keyboard('[ArrowRight]');

		const [first] = docGetAllByRole(
			projects!.parentNode as HTMLElement,
			'menuitem'
		);

		expect(first).toEqual(document.activeElement);
	});

	it('move focus to parent if focus is on child when pressing left arrow key', () => {
		const {getAllByRole} = render(
			<ClayVerticalNav items={ITEMS} spritemap="/path/to" />
		);

		const [, , projects] = getAllByRole('button');

		userEvent.tab();
		userEvent.tab();

		expect(document.activeElement).toEqual(projects);

		userEvent.keyboard('[ArrowRight]');

		expect(projects!.getAttribute('aria-expanded')).toBe('true');

		userEvent.keyboard('[ArrowRight]');

		const [first] = docGetAllByRole(
			projects!.parentNode as HTMLElement,
			'menuitem'
		);

		expect(first).toEqual(document.activeElement);

		userEvent.keyboard('[ArrowLeft]');

		expect(projects).toEqual(document.activeElement);
	});
});
