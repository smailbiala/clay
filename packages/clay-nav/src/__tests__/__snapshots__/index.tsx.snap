// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayVerticalNav renders 1`] = `
<div>
  <nav
    class="menubar menubar-transparent menubar-vertical-expand-md"
  >
    <button
      class="menubar-toggler btn btn-unstyled"
      type="button"
    >
      <span
        class="inline-item inline-item-before"
      >
        Menu
      </span>
      <svg
        class="lexicon-icon lexicon-icon-caret-bottom"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="/path/to#caret-bottom"
        />
      </svg>
    </button>
    <div
      class="collapse menubar-collapse"
    >
      <ul
        aria-orientation="vertical"
        class="nav nav-nested"
        role="menubar"
      >
        <li
          class="nav-item"
          role="none"
        >
          <button
            aria-expanded="true"
            aria-haspopup="true"
            class="nav-link collapse-icon btn btn-unstyled"
            role="button"
            type="button"
          >
            Home
            <span
              class="collapse-icon-closed"
            >
              <svg
                class="lexicon-icon lexicon-icon-caret-right"
                focusable="false"
                role="presentation"
              >
                <use
                  xlink:href="/path/to#caret-right"
                />
              </svg>
            </span>
            <span
              class="collapse-icon-open"
            >
              <svg
                class="lexicon-icon lexicon-icon-caret-bottom"
                focusable="false"
                role="presentation"
              >
                <use
                  xlink:href="/path/to#caret-bottom"
                />
              </svg>
            </span>
          </button>
          <ul
            class="nav nav-stacked"
            role="menu"
          >
            <li
              class="nav-item"
              role="none"
            >
              <a
                class="nav-link collapsed"
                href="#nested1"
                role="menuitem"
                tabindex="-1"
              >
                Nested1
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
